runtime/compiler/net/MessageBuffer.hpp

```diff
@@ -0,0 +1,199 @@
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef MESSAGE_BUFFER_H
+#define MESSAGE_BUFFER_H
+
+#include "env/jittypes.h"
+#include "env/TRMemory.hpp"
+
+namespace JITServer
+{
+/**
+   @class MessageBuffer
+   @brief A wrapper around a contiguous, persistent memory allocated buffer
+   for storing a JITServer message.
+
+   The buffer is extensible, i.e. when the current capacity is reached, a new,
+   larger buffer can be allocated and data copied there.
+   Since reallocation causes addresses of values inside the buffer to change, read/write operations
+   return an offset into the buffer to indicate the location of data, instead of pointers.
+
+   Method getValueAtOffset returns a pointer to data at a given offset, but be mindful that
+   the pointer might become invalid if more data is added to the buffer.
+
+   Variable _curPtr defines the boundary of the current data. Reading/writing to/from buffer
+   will always advance the pointer.
+ */
+class MessageBuffer
+   {
+public:
+   MessageBuffer();
+
+   ~MessageBuffer()
+      {
+      TR_Memory::jitPersistentFree(_storage);
+      }
+
+
+   /**
+      @brief Get the current active size of the buffer.
+
+      Note: this returns the number of bytes written to the buffer so far,
+      NOT the overall capacity of the buffer. Capacity is the number of
+      bytes allocated, but not necessarily used.
+
+      @return the size of the buffer
+   */
+   uint32_t size() const { return _curPtr - _storage; }
+
+   char *getBufferStart() const { return _storage; }
+
+   /**
+      @brief Return a pointer to the value at given offset inside the buffer.
+
+      Given a type and offset, returns the pointer of that type.
+      Behavior is only defined if offset does not exceed populated buffer size.
+
+      @return ponter of the specified type at given offset
+   */
+   template <typename T>
+   T *getValueAtOffset(uint32_t offset) const
+      {
+      TR_ASSERT_FATAL(offset < size(), "Offset is outside of buffer bounds");
+      return reinterpret_cast<T *>(_storage + offset);
+      }
+
+   /**
+      @brief Write value of type T to the buffer.
+
+      Copies the value into buffer, expanding it if needed,
+      and advances _curPtr by sizeof(T) bytes plus some padding
+      so that the data inside the buffer is always 32-bit aligned
+      Behavior is undefined if T is not trivially copyable (i.e. not contiguous in memory).
+
+      @param val value to be written
+
+      @return offset to the beginning of written value inside the buffer
+   */
+   template <typename T>
+   uint32_t writeValue(const T &val)
+      {
+      uint8_t paddingSize = static_cast<uint8_t>(((sizeof(T) + 3) & 0xfffffffc) - sizeof(T));
```

Comment:
Can we use symbols for the `3` and `0xfffffffc` and add `static_assert(std::is_trivially_copyable<T>::value == true, "T must be trivially copyable.");`?

Reply:
Done